File: extension/tests/sandbox.test.ts
import { describe, it, expect } from 'vitest';
import fs from 'fs';
import path from 'path';
import { deterministicHex, sandboxValue, sandboxEnv } from '../src/sandbox';
import { parse } from '../src/envparser';
import { detect } from '../src/detector';

const TESTDATA = path.resolve(__dirname, '../../testdata');

describe('sandbox', () => {
    it('deterministicHex produces correct length', () => {
        const h = deterministicHex('proj123', 'MY_KEY', 20);
        expect(h.length).toBe(20);
        expect(h).toMatch(/^[0-9a-f]+$/);
    });

    it('deterministicHex produces correct length when requesting short hex', () => {
        const h = deterministicHex('proj123', 'MY_KEY', 12);
        expect(h.length).toBe(12);
    });

    it('deterministicHex is deterministic (same inputs → same output)', () => {
        const h1 = deterministicHex('proj', 'key', 16);
        const h2 = deterministicHex('proj', 'key', 16);
        expect(h1).toBe(h2);
    });

    it('deterministicHex differs by project hash', () => {
        const h1 = deterministicHex('hash_a', 'key', 16);
        const h2 = deterministicHex('hash_b', 'key', 16);
        expect(h1).not.toBe(h2);
    });

    it('deterministicHex differs by key', () => {
        const h1 = deterministicHex('proj', 'KEY_A', 16);
        const h2 = deterministicHex('proj', 'KEY_B', 16);
        expect(h1).not.toBe(h2);
    });

    it('Stripe test key passthrough (sk_test_ → unchanged)', () => {
        const result = sandboxValue('STRIPE_KEY', 'sk_test_xxx123', 'StripeKey', 'proj');
        expect(result).toBe('sk_test_xxx123');
    });

    it('Stripe test key passthrough (pk_test_ → unchanged)', () => {
        const result = sandboxValue('STRIPE_PK', 'pk_test_yyy456', 'StripeKey', 'proj');
        expect(result).toBe('pk_test_yyy456');
    });

    it('Stripe live sk_ key → sk_test_cloak_sandbox_000000000000', () => {
        const result = sandboxValue('STRIPE_SK', 'sk_test_FAKE4eC39HqLyjWDarjtT1zd', 'StripeKey', 'proj');
        expect(result).toBe('sk_test_cloak_sandbox_000000000000');
    });

    it('Stripe live pk_ key → pk_test_cloak_sandbox_000000000000', () => {
        const result = sandboxValue('STRIPE_PK', 'pk_test_FAKEoMQauvdEDq54NiTphI7jx', 'StripeKey', 'proj');
        expect(result).toBe('pk_test_cloak_sandbox_000000000000');
    });

    it('database URL sandbox values: postgres', () => {
        const result = sandboxValue('DATABASE_URL', 'postgres://admin:s3cret@db.prod.com:5432/myapp', 'DatabaseUrl', 'proj');
        expect(result).toBe('postgres://dev:dev@localhost:5432/devdb');
    });

    it('database URL sandbox values: mysql', () => {
        const result = sandboxValue('DB_URL', 'mysql://user:pass@host:3306/mydb', 'DatabaseUrl', 'proj');
        expect(result).toBe('mysql://dev:dev@localhost:3306/devdb');
    });

    it('database URL sandbox values: mongodb', () => {
        const result = sandboxValue('MONGO_URI', 'mongodb://user:pass@host:27017/db', 'DatabaseUrl', 'proj');
        expect(result).toBe('mongodb://localhost:27017/devdb');
    });

    it('database URL sandbox values: redis', () => {
        const result = sandboxValue('REDIS_URL', 'redis://:pass@cache.prod.com:6379/0', 'DatabaseUrl', 'proj');
        expect(result).toBe('redis://localhost:6379');
    });

    it('AWS access key → AKIAIOSFODNN7EXAMPLE', () => {
        const result = sandboxValue('AWS_ACCESS_KEY_ID', 'AKIAIOSFODNN7REAL123', 'AwsAccessKey', 'proj');
        expect(result).toBe('AKIAIOSFODNN7EXAMPLE');
    });

    it('AWS secret key → standard example key', () => {
        const result = sandboxValue('AWS_SECRET_ACCESS_KEY', 'real-secret', 'AwsSecretKey', 'proj');
        expect(result).toBe('wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY');
    });

    it('JWT secret → cloak-dev-jwt-secret-not-real-000000', () => {
        const result = sandboxValue('JWT_SECRET', 'real-jwt-secret', 'JwtSecret', 'proj');
        expect(result).toBe('cloak-dev-jwt-secret-not-real-000000');
    });

    it('ApiKey (sk- prefix) → sk-cloak-sandbox-key-{hex20}', () => {
        const result = sandboxValue('OPENAI_KEY', 'sk-' + 'a'.repeat(40), 'ApiKey', 'proj');
        expect(result).toMatch(/^sk-cloak-sandbox-key-[0-9a-f]{20}$/);
    });

    it('ApiKey (non-sk) → cloak_sandbox_api_key_{hex20}', () => {
        const result = sandboxValue('API_KEY', 'SG.xxxxx.yyyyyyy', 'ApiKey', 'proj');
        expect(result).toMatch(/^cloak_sandbox_api_key_[0-9a-f]{20}$/);
    });

    it('PrivateKey → cloak-sandbox-private-key-not-real', () => {
        const result = sandboxValue('PRIVATE_KEY', '-----BEGIN RSA PRIVATE KEY-----', 'PrivateKey', 'proj');
        expect(result).toBe('cloak-sandbox-private-key-not-real');
    });

    it('Password → cloak_sandbox_password', () => {
        const result = sandboxValue('DB_PASSWORD', 'real_password', 'Password', 'proj');
        expect(result).toBe('cloak_sandbox_password');
    });

    it('Token (ghp_) → ghp_cloaksandbox{hex20}', () => {
        const result = sandboxValue('GITHUB_TOKEN', 'ghp_FAKE0000000000000000000000000000000000', 'Token', 'proj');
        expect(result).toMatch(/^ghp_cloaksandbox[0-9a-f]{20}$/);
    });

    it('Token (xoxb-) → xoxb-0000-0000-cloaksandboxtoken', () => {
        const result = sandboxValue('SLACK_TOKEN', 'xoxb-123-456-abc', 'Token', 'proj');
        expect(result).toBe('xoxb-0000-0000-cloaksandboxtoken');
    });

    it('Token (eyJ JWT) → eyJjbG9hayI6InNhbmRib3gifQ==', () => {
        const result = sandboxValue('AUTH_TOKEN', 'eyJhbGciOiJIUzI1NiJ9.xxx.yyy', 'Token', 'proj');
        expect(result).toBe('eyJjbG9hayI6InNhbmRib3gifQ==');
    });

    it('Token (generic) → cloak_sandbox_token_{hex20}', () => {
        const result = sandboxValue('SESSION_TOKEN', 'some-session-token-value', 'Token', 'proj');
        expect(result).toMatch(/^cloak_sandbox_token_[0-9a-f]{20}$/);
    });

    it('Unknown → cloak_sandbox_{hex20}', () => {
        const result = sandboxValue('MYSTERY', 'highentropy123', 'Unknown', 'proj');
        expect(result).toMatch(/^cloak_sandbox_[0-9a-f]{20}$/);
    });

    it('GenericUrl → replaces credentials with dev:dev@', () => {
        const result = sandboxValue('SOME_URL', 'https://user:pass@example.com/path', 'GenericUrl', 'proj');
        expect(result).toBe('https://dev:dev@example.com/path');
    });

    it('sandboxEnv preserves non-secrets', () => {
        const content = 'NODE_ENV=production\nPORT=3000\nDEBUG=false\n';
        const result = sandboxEnv(content, 'proj123');
        expect(result).toBe(content);
    });

    it('sandboxEnv replaces DATABASE_URL', () => {
        const content = 'DATABASE_URL=postgres://admin:s3cret@db.prod.com:5432/myapp\nNODE_ENV=production\n';
        const result = sandboxEnv(content, 'proj123');
        expect(result).toContain('DATABASE_URL=postgres://dev:dev@localhost:5432/devdb');
        expect(result).toContain('NODE_ENV=production');
    });

    it('sandboxEnv preserves comments and blank lines', () => {
        const content = '# This is a comment\n\nDB_PASSWORD=secret123\n';
        const result = sandboxEnv(content, 'proj123');
        expect(result).toContain('# This is a comment');
        expect(result).toContain('\n\n');
        expect(result).toContain('DB_PASSWORD=cloak_sandbox_password');
    });

    it('sandboxEnv uses consistent sandbox values (deterministic)', () => {
        const content = 'MY_SECRET=aB3xY7mNqR2vZ8wL5kP9cT1uE4sD6jF0gH\n';
        const result1 = sandboxEnv(content, 'myhash');
        const result2 = sandboxEnv(content, 'myhash');
        expect(result1).toBe(result2);
    });

    it('sandboxEnv: different project hashes produce different Unknown sandbox values', () => {
        const content = 'SOME_HASH=aB3xY7mNqR2vZ8wL5kP9cT1uE4sD6jF0gH\n';
        const result1 = sandboxEnv(content, 'hash_a');
        const result2 = sandboxEnv(content, 'hash_b');
        expect(result1).not.toBe(result2);
    });
});

File: cli/tests/sandbox_test.rs
//! Integration tests for the sandbox module.

use cloak::detector::SecretType;
use cloak::sandbox::{deterministic_hex, sandbox_env, sandbox_value};

const PROJ: &str = "testproject123";

// ── 1. sandbox_database_url_postgres ─────────────────────────────────────────

#[test]
fn sandbox_database_url_postgres() {
    let result = sandbox_value(
        "DATABASE_URL",
        "postgres://user:secret@db.example.com:5432/myapp",
        &SecretType::DatabaseUrl,
        PROJ,
    );
    assert_eq!(result, "postgres://dev:dev@localhost:5432/devdb");
    println!("postgres sandbox: {}", result);
}

// ── 2. sandbox_database_url_redis ────────────────────────────────────────────

#[test]
fn sandbox_database_url_redis() {
    let result = sandbox_value(
        "REDIS_URL",
        "redis://:password@redis.example.com:6379",
        &SecretType::DatabaseUrl,
        PROJ,
    );
    assert_eq!(result, "redis://localhost:6379");
    println!("redis sandbox: {}", result);
}

// ── 3. sandbox_stripe_live_key ───────────────────────────────────────────────

#[test]
fn sandbox_stripe_live_key() {
    let result = sandbox_value(
        "STRIPE_SECRET_KEY",
        "sk_test_FAKE23xyz456def789ghi012jkl345",
        &SecretType::StripeKey,
        PROJ,
    );
    assert_eq!(result, "sk_test_cloak_sandbox_000000000000");
    println!("stripe live sandbox: {}", result);

    // pk_live_ also replaced
    let pk = sandbox_value(
        "STRIPE_PK",
        "pk_test_FAKE23xyz456def789ghi012jkl345",
        &SecretType::StripeKey,
        PROJ,
    );
    assert_eq!(pk, "pk_test_cloak_sandbox_000000000000");
}

// ── 4. sandbox_stripe_test_key_passthrough ───────────────────────────────────

#[test]
fn sandbox_stripe_test_key_passthrough() {
    let sk_test = "sk_test_abc123xyz456def789ghi012jkl345";
    let result = sandbox_value("STRIPE_SECRET_KEY", sk_test, &SecretType::StripeKey, PROJ);
    assert_eq!(result, sk_test, "sk_test_ must pass through unchanged");

    let pk_test = "pk_test_abc123xyz456def789ghi012jkl345";
    let result_pk = sandbox_value("STRIPE_PK", pk_test, &SecretType::StripeKey, PROJ);
    assert_eq!(result_pk, pk_test, "pk_test_ must pass through unchanged");

    println!("stripe test passthrough OK");
}

// ── 5. sandbox_aws_keys ───────────────────────────────────────────────────────

#[test]
fn sandbox_aws_keys() {
    let access = sandbox_value(
        "AWS_ACCESS_KEY_ID",
        "AKIAREALKEY1234567890",
        &SecretType::AwsAccessKey,
        PROJ,
    );
    assert_eq!(access, "AKIAIOSFODNN7EXAMPLE");

    let secret = sandbox_value(
        "AWS_SECRET_ACCESS_KEY",
        "realSecretKeyHere",
        &SecretType::AwsSecretKey,
        PROJ,
    );
    assert_eq!(secret, "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY");

    println!("aws sandbox access: {}", access);
    println!("aws sandbox secret: {}", secret);
}

// ── 6. sandbox_jwt_secret ────────────────────────────────────────────────────

#[test]
fn sandbox_jwt_secret() {
    let result = sandbox_value(
        "JWT_SECRET",
        "my-super-secret-jwt-signing-key-very-long",
        &SecretType::JwtSecret,
        PROJ,
    );
    assert_eq!(result, "cloak-dev-jwt-secret-not-real-000000");
    println!("jwt sandbox: {}", result);
}

// ── 7. sandbox_api_key ───────────────────────────────────────────────────────

#[test]
fn sandbox_api_key() {
    // Generic API key (non sk- prefix)
    let result = sandbox_value(
        "SOME_API_KEY",
        "some-random-api-key-value-1234567890",
        &SecretType::ApiKey,
        PROJ,
    );
    assert!(
        result.starts_with("cloak_sandbox_api_key_"),
        "expected cloak_sandbox_api_key_ prefix, got: {}",
        result
    );

    // sk- prefixed API key
    let sk_result = sandbox_value(
        "OPENAI_API_KEY",
        "sk-abc123def456ghi789jkl012mno345pqr678stu901vwx234y",
        &SecretType::ApiKey,
        PROJ,
    );
    assert!(
        sk_result.starts_with("sk-cloak-sandbox-key-"),
        "expected sk-cloak-sandbox-key- prefix, got: {}",
        sk_result
    );

    println!("api key sandbox: {}", result);
    println!("openai api key sandbox: {}", sk_result);
}

// ── 8. sandbox_token_github ──────────────────────────────────────────────────

#[test]
fn sandbox_token_github() {
    let result = sandbox_value(
        "GITHUB_TOKEN",
        "ghp_FAKE0000000000000000000000000000000000",
        &SecretType::Token,
        PROJ,
    );
    assert!(
        result.starts_with("ghp_cloaksandbox"),
        "expected ghp_cloaksandbox prefix, got: {}",
        result
    );
    println!("github token sandbox: {}", result);

    // Slack token
    let slack = sandbox_value(
        "SLACK_BOT_TOKEN",
        "xoxb-0000-0000-FAKEefghijklmnop",
        &SecretType::Token,
        PROJ,
    );
    assert_eq!(slack, "xoxb-0000-0000-cloaksandboxtoken");

    // JWT token
    let jwt = sandbox_value(
        "ACCESS_TOKEN",
        "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.payload.sig",
        &SecretType::Token,
        PROJ,
    );
    assert_eq!(jwt, "eyJjbG9hayI6InNhbmRib3gifQ==");
}

// ── 9. sandbox_deterministic ─────────────────────────────────────────────────

#[test]
fn sandbox_deterministic() {
    let key = "MY_SECRET_KEY";
    let value = "some-high-entropy-secret-value-xyz";
    let secret_type = SecretType::Unknown;

    // Same inputs → same output (called twice)
    let r1 = sandbox_value(key, value, &secret_type, "hash_abc");
    let r2 = sandbox_value(key, value, &secret_type, "hash_abc");
    assert_eq!(r1, r2, "same inputs must produce same output");

    // Different project hash → different output
    let r3 = sandbox_value(key, value, &secret_type, "hash_xyz");
    assert_ne!(
        r1, r3,
        "different project hash must produce different output"
    );

    // deterministic_hex helper itself is deterministic
    let h1 = deterministic_hex("proj", "key", 32);
    let h2 = deterministic_hex("proj", "key", 32);
    assert_eq!(h1, h2);

    // deterministic_hex differs by project_hash
    let h3 = deterministic_hex("other", "key", 32);
    assert_ne!(h1, h3);

    println!("deterministic test passed: {} vs {}", r1, r3);
}

// ── 10. sandbox_env_preserves_structure ──────────────────────────────────────

#[test]
fn sandbox_env_preserves_structure() {
    let content = std::fs::read_to_string(
        std::path::Path::new(env!("CARGO_MANIFEST_DIR")).join("tests/testdata/realistic.env"),
    )
    .expect("failed to read realistic.env");

    let result = sandbox_env(&content, PROJ);

    println!("=== sandbox_env output ===");
    println!("{}", result);
    println!("==========================");

    // Comments must be preserved
    assert!(
        result.contains("# Application configuration"),
        "comment '# Application configuration' must be preserved"
    );
    assert!(
        result.contains("# Database"),
        "comment '# Database' must be preserved"
    );
    assert!(
        result.contains("# Stripe"),
        "comment '# Stripe' must be preserved"
    );

    // Non-secrets must be unchanged
    assert!(
        result.contains("NODE_ENV=development"),
        "NODE_ENV=development must be unchanged"
    );
    assert!(result.contains("PORT=3000"), "PORT=3000 must be unchanged");
    assert!(
        result.contains("HOST=localhost"),
        "HOST=localhost must be unchanged"
    );
    assert!(
        result.contains("APP_NAME=MyApplication"),
        "APP_NAME=MyApplication must be unchanged"
    );
    assert!(
        result.contains("LOG_LEVEL=info"),
        "LOG_LEVEL=info must be unchanged"
    );
    assert!(
        result.contains("DEBUG=false"),
        "DEBUG=false must be unchanged"
    );

    // Real secrets must be replaced — original values must not appear
    assert!(
        !result.contains("postgres://user:secret@db.example.com"),
        "real postgres URL must not appear in output"
    );
    assert!(
        !result.contains("sk_live_abc123"),
        "real Stripe live key must not appear in output"
    );
    assert!(
        !result.contains("AKIAREALKEY"),
        // The test file uses AKIAIOSFODNN7EXAMPLE which is already the sandbox value
        "real AWS key must not appear"
    );

    // Sandbox replacements must be present
    assert!(
        result.contains("postgres://dev:dev@localhost:5432/devdb"),
        "postgres sandbox value must appear"
    );
    assert!(
        result.contains("sk_test_cloak_sandbox_000000000000"),
        "stripe sandbox value must appear"
    );

    // sk_test_ must pass through unchanged
    assert!(
        result.contains("STRIPE_TEST_KEY=sk_test_abc123xyz456def789ghi012jkl345"),
        "sk_test_ key must pass through unchanged"
    );

    // Structure: result should have roughly same number of lines
    let original_lines = content.lines().count();
    let result_lines = result.lines().count();
    assert_eq!(
        original_lines, result_lines,
        "line count must be preserved: original={}, result={}",
        original_lines, result_lines
    );
}
